src/compiler/checker.ts

```diff
@@ -19381,7 +19381,24 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
         for (let i = 0; i < paramCount; i++) {
             const sourceType = i === restIndex ? getRestTypeAtPosition(source, i) : tryGetTypeAtPosition(source, i);
-            const targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            let targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            if (i === restIndex && targetType && sourceType && isTupleType(sourceType)) {

```

Comment:
I think I have a reasonable understanding now, so feel free to ignore the question, but it might still be helpful to have a sort of high level description.

Reply:
I added source comments in which I try to explain what is happening here, let me know what do you thnk about them.